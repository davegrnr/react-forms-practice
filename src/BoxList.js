import React, { useState } from 'react';
import NewBoxForm from './NewBoxForm'
import Box from './Box'
import { v4 as uuid } from 'uuid'

const BoxList = () => {
    const INITIAL_STATE = []

    const [boxes, setBoxes] = useState(INITIAL_STATE)

    const addBox = (newBox) => {
        setBoxes(boxes => [...boxes, {...newBox, id: uuid()}])
    }
    const removeBox = id => {
        setBoxes(boxes => boxes.filter(box => box.id !== id));
      };

    return (
        <div>
            <h3>Create a New Box!</h3>
            <NewBoxForm addBox={addBox} />

            <div>
                {boxes.map(({id, color, height, width}) => 
                    <Box 
                        id={id} 
                        color={color} 
                        height={height} 
                        width={width}
                        key={id} 
                        handleRemove={removeBox}
                    />)}
            </div>
        </div>
    )
}

export default BoxList