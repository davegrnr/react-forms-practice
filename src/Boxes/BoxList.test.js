import React from "react";
import { fireEvent, render } from "@testing-library/react";
import BoxList from "./BoxList";

function addBox(boxList, height="50", width="50", color="red"){
    const heightInput = boxList.getByLabelText("Height")
    const widthInput = boxList.getByLabelText("Width")
    const colorInput = boxList.getByLabelText("Color")
    fireEvent.change(heightInput, {target: {value: height}})
    fireEvent.change(widthInput, {target: {value: width}})
    fireEvent.change(colorInput, {target: {value: color}})
    const btn = boxList.getByText("Box Time!")
    fireEvent.click(btn)
}

it("renders without crashing", () => {
    render(<BoxList />);
});

it("matches snapshot", function() {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

it("adds a new box", () => {
    const boxList = render(<BoxList />)

    // empty box list
    expect(boxList.queryByText("Remove")).not.toBeInTheDocument()

    addBox(boxList);

    // expect remove button and box to be on the page
    const btn = boxList.getByText("Remove");
    expect(btn).toBeInTheDocument();
    expect(btn.previousSibling).toBeInTheDocument();

    expect(boxList.getAllByDisplayValue("")).toHaveLength(3)
})

it("removes a box", () => {
    const boxList = render(<BoxList />);
    addBox(boxList);

    const btn = boxList.getByText("Remove");

    fireEvent.click(btn);
    expect(btn).not.toBeInTheDocument()
})