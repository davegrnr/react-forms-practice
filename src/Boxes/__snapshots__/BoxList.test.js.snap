// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3>
      Create a New Box!
    </h3>
    <form>
      <label
        for="color"
      >
        Color 
      </label>
      <input
        id="color"
        name="color"
        placeholder="Box Color"
        type="text"
        value=""
      />
      <label
        for="height"
      >
         Height 
      </label>
      <input
        id="height"
        name="height"
        placeholder="Box Height"
        type="text"
        value=""
      />
      <label
        for="width"
      >
         Width 
      </label>
      <input
        id="width"
        name="width"
        placeholder="Box Width"
        type="text"
        value=""
      />
      <br />
      <button
        style="margin: 10px;"
      >
        Box Time!
      </button>
    </form>
    <div />
  </div>
</DocumentFragment>
`;
